libraries/AP_InertialSensor/AP_InertialSensor_SCH16T.cpp

```diff
@@ -0,0 +1,472 @@
+        !_imu.register_gyro(gyro_instance, expected_sample_rate_hz,   dev->get_bus_id_devtype(DEVTYPE_INS_SCH16T))) {
+        return;
+    }
+
+    // setup sensor rotations from probe()
+    set_gyro_orientation(gyro_instance, rotation);
+    set_accel_orientation(accel_instance, rotation);
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_InertialSensor_SCH16T::loop, void),
+                                      "SCH16T",
+                                      1024, AP_HAL::Scheduler::PRIORITY_BOOST, 1)) {
+        AP_HAL::panic("Failed to create SCH16T thread");
+    }
+}
+
+
+bool AP_InertialSensor_SCH16T::init()
+{
+    hal.console->printf("AP_InertialSensor_SCH16T::init()");
+    WITH_SEMAPHORE(dev->get_semaphore());
+
+    dev->set_speed(AP_HAL::Device::SPEED_LOW);
+
+    if (!ReadProductID()) {
+        hal.console->printf("reading product ID failed");
+        return false;
+    }
+
+    ConfigureRegisters();
+
+    return true;
+}
+
+void AP_InertialSensor_SCH16T::loop()
+{
+    while (true) {
+        uint32_t wait_us = 0;
+        bool wait_ok = true;
+
+        switch (_state) {
+        case State::Reset: {
+                failure_count = 0;
+                hal.console->printf("Resetting (soft)");
+                RegisterWrite(CTRL_RESET, SPI_SOFT_RESET);
+                _state = State::Configure;
+                wait_us = POWER_ON_TIME;
+                break;
+            }
+
+
+        case State::Configure: {
+                ConfigureRegisters();
+                _state = State::LockConfiguration;
+                wait_us = POWER_ON_TIME;
+                break;
+            }
+
+        case State::LockConfiguration: {
+                hal.console->printf("locking configuration");
+                ReadStatusRegisters(); // Read all status registers once
+                RegisterWrite(CTRL_MODE, (EOI | EN_SENSOR)); // Write EOI and EN_SENSOR
+
+                _state = State::Validate;
+                wait_us = 5000UL;
+                break;
+            }
+
+        case State::Validate: {
+                ReadStatusRegisters(); // Read all status registers twice
+                ReadStatusRegisters();
+
+                // Check that registers are configured properly and that the sensor status is OK
+                if (ValidateSensorStatus() && ValidateRegisterConfiguration()) {
+                    hal.console->printf("Starting read loop");
+                    _state = State::Read;
+                    wait_us = 1000000UL / expected_sample_rate_hz;
+                    dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+
+                } else {
+                    hal.console->printf("Sensor or Register validation failed, resetting");
+                    _state = State::Reset;
+                    wait_us = 100000UL;
+                }
+
+                break;
+            }
+
+        case State::Read: {
+                uint32_t tstart = AP_HAL::micros();
+
+                if (drdy_pin != 0) {
```

Comment:
This is usually set in hwdef so that it doesn't consume flash if you are not using it

Reply:
is there an example somewhere? I see other imu drivers doing this